src/GraphQL/DI/GraphQLBuilderBase.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using GraphQL.Caching;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQL.Validation.Complexity;
+
+namespace GraphQL.DI
+{
+    /// <summary>
+    /// Base implementation of <see cref="IGraphQLBuilder"/>.
+    /// </summary>
+    public abstract class GraphQLBuilderBase : IGraphQLBuilder
+    {
+        /// <summary>
+        /// Register the default services required by GraphQL if they have not already been registered.
+        /// <br/><br/>
+        /// Does not include <see cref="IDocumentWriter"/>, and the default <see cref="IDocumentExecuter"/>
+        /// implementation does not support subscriptions.
+        /// </summary>
+        protected virtual void Initialize()
+        {
+            // configure an error to be displayed when no IDocumentWriter is registered
+            this.TryRegister<IDocumentWriter>(_ =>
+            {
+                throw new InvalidOperationException(
+                    "IDocumentWriter not set in DI container. " +
+                    "Add a IDocumentWriter implementation, for example " +
+                    "GraphQL.SystemTextJson.DocumentWriter or GraphQL.NewtonsoftJson.DocumentWriter. " +
+                    "For more information, see: https://github.com/graphql-dotnet/graphql-dotnet/blob/master/README.md and https://github.com/graphql-dotnet/server/blob/develop/README.md.");
+            }, ServiceLifetime.Transient);
+
+            // configure service implementations to use the configured default services when not overridden by a user
+            this.TryRegister<IDocumentExecuter, DocumentExecuter>(ServiceLifetime.Singleton);
+            this.TryRegister<IDocumentBuilder, GraphQLDocumentBuilder>(ServiceLifetime.Singleton);
+            this.TryRegister<IDocumentValidator, DocumentValidator>(ServiceLifetime.Singleton);
+            this.TryRegister<IComplexityAnalyzer, ComplexityAnalyzer>(ServiceLifetime.Singleton);
+            this.TryRegister<IDocumentCache>(DefaultDocumentCache.Instance);
```

Comment:
If we register the already created instance, the DI container will not dispose it even if it supports `IDisposable`, right? So each time we should make sure. 
What about
```
 this.TryRegister<IDocumentCache>(provider => DefaultDocumentCache.Instance, ServiceLifetime.Singleton);
```
? In particular case it does not matter since `DefaultDocumentCache` does not implement `IDisposable`. I ask in general.

Reply:
Perhaps we should add test to GraphQL.MicrosoftDI to be sure instance registrations are not disposed when the service collection is disposed

Reply:
Or we can just make this create a new instance, to be on the safe side in case of other behavior by any DI provider

Reply:
Wait the default document cache = no cache. It’s not disposable.

Reply:
Oh you said that, I missed it. Ok well then I’m not sure what question I’m answering.

Reply:
I want to be sure that any disposable instance supplied by caller will not be disposed.

Reply:
Added tests in #2781 for the MS DI provider.  There are already plenty of tests that ensure that the call to `Register` calls the proper underlying method on `IServiceCollection`.